README.md
About
=====

*This is the [Rust](https://www.rust-lang.org/) version of
[frotate](https://github.com/ip1981/frotate) ;)*


`frotate` stands for "functional rotate", whatever.
This is an evolution of the [log2rotate's ideas](http://jekor.com/log2rotate).
See also [pylog2rotate](https://github.com/avian2/pylog2rotate).

`frotate` is designed to rotate backups with any balance between retention
and space usage.  Instead of rotating backups using some familiar method such
as daily, weekly, monthly, and yearly periods, it rotates backups using any
periods. Thus "functional".

The idea is simple, the rotation schedule is determined by an integer function.
This function gives us a period (number) of days when we must encounter at
least one backup or whatever we are rotating. When we use an exponential
function, the scheme is similar to the radioactive decay law. When the
function is simply a constant 1, we don't rotate anything and retain all
the backups. If it is 2, we retain each second backup.  With some trivial
function we can achieve a well-known dayly-weekly-monthly-yearly scheme.

The `frotate` command line utility implements only exponential periods with
arbitrary base (ensure it is > 1, or have fun otherwise).


Usage
=====

Note that when neither `--keep` nor `--delete` option is given, the utility
prints all intervals with all days _to standard error_ and exits with non-zero
code. In production you will need to specify `--keep` or `--delete` explicitly.

```
Usage: frotate [-k|-d] [-b <base>] <day>...
       frotate --help

Options:
  -k --keep         Print days to keep
  -d --delete       Print days to delete
  -b --base <base>  Base of the exponent [default: 1.1]
  -h --help         Show this help text

```


Example
=======

Different modes with the same days:

```
$ frotate --base 2 2019-08-31 2019-08-30 2019-08-29 2019-08-28 2019-08-27 2019-08-26 2019-08-25 2019-08-24
2019-08-31
2019-08-29 2019-08-30
2019-08-25 2019-08-26 2019-08-27 2019-08-28
2019-08-24

$ frotate --keep --base 2 2019-08-31 2019-08-30 2019-08-29 2019-08-28 2019-08-27 2019-08-26 2019-08-25 2019-08-24
2019-08-31 2019-08-29 2019-08-25 2019-08-24

$ frotate --delete --base 2 2019-08-31 2019-08-30 2019-08-29 2019-08-28 2019-08-27 2019-08-26 2019-08-25 2019-08-24
2019-08-30 2019-08-26 2019-08-27 2019-08-28
```

More or less realistic example when we keep some backups and get new ones, but not every day:

```
$ frotate --keep --base 2 2019-09-01 2019-08-31 2019-08-30 2019-08-28 2019-08-24
2019-09-01 2019-08-30 2019-08-28 2019-08-24

$ frotate --keep --base 2 2019-09-05 2019-09-01 2019-08-30 2019-08-28 2019-08-24
2019-09-05 2019-08-30 2019-08-24

$ frotate --keep --base 2 2019-09-06 2019-09-05 2019-08-30 2019-08-24
2019-09-06 2019-09-05 2019-08-24

$ frotate --keep --base 2 2019-09-07 2019-09-06 2019-09-05 2019-08-24
2019-09-07 2019-09-05 2019-08-24

$ frotate --keep --base 2 2019-09-08 2019-09-07 2019-09-06 2019-08-24
2019-09-08 2019-09-06 2019-08-24

$ frotate --keep --base 2 2019-09-09 2019-09-08 2019-09-06 2019-08-24
2019-09-09 2019-09-08 2019-09-06 2019-08-24
```



CHANGELOG.md
0.1.1 (2019-11-15)
==================

  * Only internal changes.


0.1.0 (2019-10-05)
==================

  * First version.



src/lib.rs
extern crate chrono;

use std::collections::LinkedList;

use chrono::NaiveDate;
use chrono::Duration;

#[cfg(test)]
mod tests {

    use std::collections::LinkedList;
    use chrono::NaiveDate;

    use super::partition;
    use super::partition_days;

    macro_rules! list {
        ( $( $x:expr ),* ) => {
            {
                let mut temp_list = LinkedList::new();
                $(
                    temp_list.push_back($x);
                )*
                temp_list
            }
        };
    }

    #[test]
    fn all_const1() {
        let ins = vec![1, 2, 3, 4, 5, 6, 7, 8, 9];
        let outs = list![list![1], list![2], list![3],
                         list![4], list![5], list![6],
                         list![7], list![8], list![9]];

        assert_eq!(partition(&|_n| 1, &ins), outs);
    }

    #[test]
    fn all_const2() {
        let ins = vec![1, 2, 3, 4, 5, 6, 7, 8, 9];
        let outs = list![list![1, 2], list![3, 4],
                         list![5, 6], list![7, 8], list![9]];

        assert_eq!(partition(&|_n| 2, &ins), outs);
    }

    #[test]
    fn odds_const1() {
        let ins = vec![1, 3, 5, 7, 9, 11, 13, 15];
        let outs = list![list![1], list![], list![3], list![],
                         list![5], list![], list![7], list![],
                         list![9], list![], list![11], list![],
                         list![13], list![], list![15]];

        assert_eq!(partition(&|_n| 1, &ins), outs);
    }

    #[test]
    fn odds_const2() {
        let ins = vec![1, 3, 5, 7, 9, 11, 13, 15];
        let outs = list![list![1], list![3], list![5],
                         list![7], list![9], list![11],
                         list![13], list![15]];

        assert_eq!(partition(&|_n| 2, &ins), outs);
    }

    fn exp2(n: u32) -> i64 {
        let i: i32 = n as i32 - 1;
        2_f32.powi(i).ceil() as i64
    }

    #[test]
    fn all_exp2() {
        let ins = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];
        let outs = list![list![1], list![2, 3], list![4, 5, 6, 7],
                         list![8, 9, 10, 11, 12, 13]];

        assert_eq!(partition(&exp2, &ins), outs);
    }

    #[test]
    fn dates_exp2() {
        let d = NaiveDate::from_ymd;
        let ins = vec![d(2019, 08, 30), d(2019, 08, 31),
                       d(2019, 09, 01), d(2019, 09, 02),
                       d(2019, 09, 03), d(2019, 09, 04)];
        let outs = list![list![d(2019, 09, 04)],
                         list![d(2019, 09, 03), d(2019, 09, 02)],
                         list![d(2019, 09, 01), d(2019, 08, 31), d(2019, 08, 30)]];

        assert_eq!(partition_days(&exp2, &ins), outs);
    }
}


pub fn partition(f: &Fn(u32) -> i64, v: &Vec<i64>) -> LinkedList<LinkedList<i64>> {
    let mut term: LinkedList<i64> = LinkedList::new();
    let mut res: LinkedList<LinkedList<i64>> = LinkedList::new();
    let mut n: u32 = 1;
    let mut a: i64 = v[0];

    for &i in v.iter() {
        while i >= a + f(n) {
            res.push_back(term);
            term = LinkedList::new();
            a += f(n);
            n += 1;
        }
        term.push_back(i);
    }
    res.push_back(term);

    res
}

pub fn partition_days(
    f: &Fn(u32) -> i64,
    days: &Vec<NaiveDate>,
) -> LinkedList<LinkedList<NaiveDate>> {
    let day1 = days[0];
    let part;

    {
        let mut v: Vec<i64> = Vec::with_capacity(days.len());

        for &d in days.iter() {
            v.push((day1 - d).num_days());
        }
        v.sort_unstable();
        v.dedup();

        part = partition(f, &v);
    }

    let res = part.into_iter()
        .map(|l| {
            l.into_iter()
                .map(|d| day1 - Duration::days(d))
                .collect()
        })
        .collect();

    res
}


src/main.rs
extern crate chrono;
extern crate docopt;
extern crate serde;

use std::process::exit;

use chrono::NaiveDate;
use docopt::Docopt;
use serde::Deserialize;
use std::collections::LinkedList;

mod partition;
use partition::partition_days;

const USAGE: &'static str = "
Usage: frotate [-k|-d] [-b <base>] <day>...
       frotate --help

Options:
  -k --keep         Print days to keep
  -d --delete       Print days to delete
  -b --base <base>  Base of the exponent [default: 1.1]
  -h --help         Show this help text
";

#[derive(Deserialize)]
struct Args {
    flag_keep: bool,
    flag_delete: bool,
    flag_base: f32,
    arg_day: Vec<NaiveDate>,
}

fn exponent(b: f32, n: u32) -> i64 {
    let i: i32 = n as i32 - 1;
    b.powi(i).ceil() as i64
}

fn main() {
    let args: Args = Docopt::new(USAGE)
        .and_then(|d| d.deserialize())
        .unwrap_or_else(|e| e.exit());

    let parts: LinkedList<LinkedList<NaiveDate>> =
        partition_days(&|n| exponent(args.flag_base, n), &args.arg_day)
            .into_iter()
            .map(|l| l.into_iter().rev().collect())
            .collect();

    if args.flag_keep {
        for days in parts.iter() {
            match days.front() {
                Some(d) => print!("{:?} ", d),
                None => {}
            }
        }
        println!();
    } else if args.flag_delete {
        for days in parts.iter() {
            for d in days.iter().skip(1) {
                print!("{:?} ", d);
            }
        }
        println!();
    } else {
        for days in parts.iter() {
            for d in days.iter() {
                eprint!("{:?} ", d);
            }
            eprintln!();
        }
        exit(1);
    }
}


src/partition.rs
extern crate chrono;

use std::collections::LinkedList;

use chrono::Duration;
use chrono::NaiveDate;

#[cfg(test)]
mod tests {

    use chrono::NaiveDate;
    use std::collections::LinkedList;

    use super::partition;
    use super::partition_days;

    macro_rules! list {
        ( $( $x:expr ),* ) => {
            {
                let mut temp_list = LinkedList::new();
                $(
                    temp_list.push_back($x);
                )*
                temp_list
            }
        };
    }

    #[test]
    fn all_const1() {
        let ins = vec![1, 2, 3, 4, 5, 6, 7, 8, 9];
        let outs = list![
            list![1],
            list![2],
            list![3],
            list![4],
            list![5],
            list![6],
            list![7],
            list![8],
            list![9]
        ];

        assert_eq!(partition(&|_n| 1, &ins), outs);
    }

    #[test]
    fn all_const2() {
        let ins = vec![1, 2, 3, 4, 5, 6, 7, 8, 9];
        let outs = list![list![1, 2], list![3, 4], list![5, 6], list![7, 8], list![9]];

        assert_eq!(partition(&|_n| 2, &ins), outs);
    }

    #[test]
    fn odds_const1() {
        let ins = vec![1, 3, 5, 7, 9, 11, 13, 15];
        let outs = list![
            list![1],
            list![],
            list![3],
            list![],
            list![5],
            list![],
            list![7],
            list![],
            list![9],
            list![],
            list![11],
            list![],
            list![13],
            list![],
            list![15]
        ];

        assert_eq!(partition(&|_n| 1, &ins), outs);
    }

    #[test]
    fn odds_const2() {
        let ins = vec![1, 3, 5, 7, 9, 11, 13, 15];
        let outs = list![
            list![1],
            list![3],
            list![5],
            list![7],
            list![9],
            list![11],
            list![13],
            list![15]
        ];

        assert_eq!(partition(&|_n| 2, &ins), outs);
    }

    fn exp2(n: u32) -> i64 {
        let i: i32 = n as i32 - 1;
        2_f32.powi(i).ceil() as i64
    }

    #[test]
    fn all_exp2() {
        let ins = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];
        let outs = list![
            list![1],
            list![2, 3],
            list![4, 5, 6, 7],
            list![8, 9, 10, 11, 12, 13]
        ];

        assert_eq!(partition(&exp2, &ins), outs);
    }

    #[test]
    fn dates_exp2() {
        let d = NaiveDate::from_ymd;
        let ins = vec![
            d(2019, 08, 30),
            d(2019, 08, 31),
            d(2019, 09, 01),
            d(2019, 09, 02),
            d(2019, 09, 03),
            d(2019, 09, 04),
        ];
        let outs = list![
            list![d(2019, 09, 04)],
            list![d(2019, 09, 03), d(2019, 09, 02)],
            list![d(2019, 09, 01), d(2019, 08, 31), d(2019, 08, 30)]
        ];

        assert_eq!(partition_days(&exp2, &ins), outs);
    }
}

pub fn partition(f: &dyn Fn(u32) -> i64, v: &Vec<i64>) -> LinkedList<LinkedList<i64>> {
    let mut term: LinkedList<i64> = LinkedList::new();
    let mut res: LinkedList<LinkedList<i64>> = LinkedList::new();
    let mut n: u32 = 1;
    let mut a: i64 = v[0];

    for &i in v.iter() {
        while i >= a + f(n) {
            res.push_back(term);
            term = LinkedList::new();
            a += f(n);
            n += 1;
        }
        term.push_back(i);
    }
    res.push_back(term);

    res
}

pub fn partition_days(
    f: &dyn Fn(u32) -> i64,
    days: &Vec<NaiveDate>,
) -> LinkedList<LinkedList<NaiveDate>> {
    let day1 = days[0];
    let part;

    {
        let mut v: Vec<i64> = Vec::with_capacity(days.len());

        for &d in days.iter() {
            v.push((day1 - d).num_days());
        }
        v.sort_unstable();
        v.dedup();

        part = partition(f, &v);
    }

    let res = part
        .into_iter()
        .map(|l| l.into_iter().map(|d| day1 - Duration::days(d)).collect())
        .collect();

    res
}
